### PR TITLE
Support file editing in azure explorer

### PR DESCRIPTION
**Pull request by Flanker32**
Support file editing in azure explorer, for task [#1788688](https://dev.azure.com/mseng/VSJava/_sprints/taskboard/Java%20Azure%20Tooling/VSJava/Sprint%20180?workitem=1788688)
- Always fetch latest `AppServiceFile` in case file was deleted and update file raw info
- Subscribe `beforeFileClosed` event with `messageBusConnection` and prompt choice dialog for user to save changes
- Prompt user again if target file was deleted or modified after open

Todo:
- Refresh parent of file node if file was deleted

### CODE REVIEW

**Comment by andxu** [comment]
Please attach the UI snapshot in the description

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java [hunk 1]
```diff
@@ -75,11 +80,19 @@
...
 import java.nio.charset.StandardCharsets;
+import java.time.ZonedDateTime;
 import java.util.*;
 import java.util.concurrent.ExecutionException;
 
 @Log
 public class IDEHelperImpl implements IDEHelper {
+
+    private static final String APP_SERVICE_FILE_EDITING = "App Service File Editing";
+    private static final String FILE_HAS_BEEN_DELETED = "Target file has been deleted from Azure, do you still want to save your changes?";
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java:91]: message template?

**Comment by Flanker32** [comment]
> Please attach the UI snapshot in the description

No new UI form in this pr, we will just prompt yes no dialog after close file editor

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java [hunk 2]
```diff
@@ -359,23 +370,28 @@ public void openLinkInBrowser(@NotNull String url) {
 
     @AzureOperation(
         value = "open file[%s] in editor",
-        params = {"$file.getName()"},
+        params = {"target.getName()"},
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java:373]: `$target`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java [hunk 3]
```diff
@@ -75,11 +80,18 @@
...
 import java.nio.charset.StandardCharsets;
+import java.time.ZonedDateTime;
 import java.util.*;
 import java.util.concurrent.ExecutionException;
 
 @Log
 public class IDEHelperImpl implements IDEHelper {
+
+    private static final String APP_SERVICE_FILE_EDITING = "App Service File Editing";
+    private static final String FILE_HAS_BEEN_DELETED = "File %s has been deleted from Azure, do you still want to save your changes?";
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java:91]: from Azure -> from remote server, do you want to create a new file with the changed content?

Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java:91]: File '%s'

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java [hunk 4]
```diff
@@ -297,8 +309,7 @@ private static Project findOpenProject(@NotNull ProjectDescriptor projectDescrip
         Project project = null;
 
         for (Project openProject : ProjectManager.getInstance().getOpenProjects()) {
-            if (projectDescriptor.getName().equals(openProject.getName())
-                && projectDescriptor.getPath().equals(openProject.getBasePath())) {
+            if (projectDescriptor.getName().equals(openProject.getName()) && projectDescriptor.getPath().equals(openProject.getBasePath())) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java:312]: StringUtils.equals

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java [hunk 5]
```diff
@@ -393,6 +409,69 @@ public void openAppServiceFile(final AppServiceFile file, Object context) {
...
+    private boolean openFileInEditor(final AppServiceFile appServiceFile, VirtualFile virtualFile, FileEditorManager fileEditorManager) {
+        final FileEditor[] editors = fileEditorManager.openFile(virtualFile, true, true);
+        if (editors.length == 0) {
+            return false;
+        }
+        for (FileEditor fileEditor : editors) {
+            if (fileEditor instanceof TextEditor) {
+                final String originContent = getTextEditorContent((TextEditor) fileEditor);
+                final MessageBusConnection messageBusConnection = fileEditorManager.getProject().getMessageBus().connect(fileEditor);
+                messageBusConnection.subscribe(FileEditorManagerListener.Before.FILE_EDITOR_MANAGER, new FileEditorManagerListener.Before() {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java:421]: shall we subscribe a listener in a loop?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java [hunk 6]
```diff
@@ -393,6 +409,69 @@ public void openAppServiceFile(final AppServiceFile file, Object context) {
...
     }
 
+    private boolean openFileInEditor(final AppServiceFile appServiceFile, VirtualFile virtualFile, FileEditorManager fileEditorManager) {
+        final FileEditor[] editors = fileEditorManager.openFile(virtualFile, true, true);
+        if (editors.length == 0) {
+            return false;
+        }
+        for (FileEditor fileEditor : editors) {
+            if (fileEditor instanceof TextEditor) {
+                final String originContent = getTextEditorContent((TextEditor) fileEditor);
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java:419]: if it is for comparing content changes, use md5 is enough

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java [hunk 7]
```diff
@@ -393,6 +409,69 @@ public void openAppServiceFile(final AppServiceFile file, Object context) {
...
+        value = "save file[%s] to azure",
+        params = {"appServiceFile.getName()"},
+        type = AzureOperation.Type.SERVICE
+    )
+    private void saveFileToAzure(final AppServiceFile appServiceFile, final String content, final Project project) {
+        AzureTaskManager.getInstance().runInBackground(new AzureTask(project, String.format("Saving %s", appServiceFile.getName()), false, () -> {
+            final AppServiceFileService fileService = AppServiceFileService.forApp(appServiceFile.getApp());
+            final AppServiceFile target = fileService.getFileByPath(appServiceFile.getPath());
+            if (target == null) {
+                boolean result = DefaultLoader.getUIHelper().showYesNoDialog(null, String.format(FILE_HAS_BEEN_DELETED, appServiceFile.getName()),
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java:459]: continueToSave

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java [hunk 8]
```diff
@@ -393,6 +409,69 @@ public void openAppServiceFile(final AppServiceFile file, Object context) {
...
+                                                                             APP_SERVICE_FILE_EDITING, Messages.getQuestionIcon());
+                if (!result) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java:461]: combine the if clause with previous showYesNoDialog

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java [hunk 9]
```diff
@@ -393,6 +409,69 @@ public void openAppServiceFile(final AppServiceFile file, Object context) {
...
+        return true;
+    }
+
+    private static String getTextEditorContent(TextEditor textEditor) {
+        return textEditor.getEditor().getDocument().getText();
+    }
+
+    @AzureOperation(
+        value = "save file[%s] to azure",
+        params = {"appServiceFile.getName()"},
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java:451]: `$appServiceFile`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java [hunk 10]
```diff
@@ -393,6 +409,69 @@ public void openAppServiceFile(final AppServiceFile file, Object context) {
...
+        type = AzureOperation.Type.SERVICE
+    )
+    private void saveFileToAzure(final AppServiceFile appServiceFile, final String content, final Project project) {
+        AzureTaskManager.getInstance().runInBackground(new AzureTask(project, String.format("Saving %s", appServiceFile.getName()), false, () -> {
+            final AppServiceFileService fileService = AppServiceFileService.forApp(appServiceFile.getApp());
+            final AppServiceFile target = fileService.getFileByPath(appServiceFile.getPath());
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java:457]: try guiding your code by natural words.
```
toSave = true;
deleted = target==null;
outdated = ZonedDateTime.parse(target.getMtime()).isAfter(ZonedDateTime.parse(appServiceFile.getMtime()));
if(deleted){
    toSave = confirm();
} else if(outdated){
    toSave = confirm();
}
if(toSave){
    saveFile()
    showNotification(succeeded)
}else{
    showNotification(cancelled)
}
```

File: Utils/azure-explorer-common/src/com/microsoft/azure/toolkit/lib/appservice/file/AppServiceFileService.java [hunk 11]
```diff
@@ -52,6 +57,21 @@ private AppServiceFileService(final WebAppBase app, AppServiceFileClient client)
         this.client = client;
     }
 
+    @Nullable
+    @AzureOperation(
+            value = "get files with path[%s] of webapp[%s]",
+            params = {"path", "@app.name()"},
```
Comment (wangmingliang-ms) [inline Utils/azure-explorer-common/src/com/microsoft/azure/toolkit/lib/appservice/file/AppServiceFileService.java:63]: `$path`

**Review by wangmingliang-ms — COMMENTED** [review]
I left some comments

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java (hunk 8 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java:461]: `if(DefaultLoader.getUIHelper().showYesNoDialog(
                    null, String.format(FILE_HAS_BEEN_MODIFIED, appServiceFile.getName()), APP_SERVICE_FILE_EDITING, Messages.getQuestionIcon()))`? I think the if condition may be too long to understand

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java (hunk 5 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java:424]: Using loop as `fileEditorManager.openFile` returns `FileEditor[]`, I will check whether there is only one editor for text file and do correspond refactor in later pr. Besides, all listener is bind to correspond `FileEditor` and will be unsubscribe when editor closed.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java [hunk 12]
```diff
@@ -393,6 +409,69 @@ public void openAppServiceFile(final AppServiceFile file, Object context) {
         AzureTaskManager.getInstance().runInModal(task);
     }
 
+    private boolean openFileInEditor(final AppServiceFile appServiceFile, VirtualFile virtualFile, FileEditorManager fileEditorManager) {
+        final FileEditor[] editors = fileEditorManager.openFile(virtualFile, true, true);
+        if (editors.length == 0) {
+            return false;
+        }
+        for (FileEditor fileEditor : editors) {
```
[the remark this answers is not in the document]
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java:420]: Using loop as fileEditorManager.openFile returns FileEditor[], I will check whether there is only one editor for text file and do correspond refactor in later pr.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java (hunk 6 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/IDEHelperImpl.java:422]: Will do in later pr.

**Review by wangmingliang-ms — APPROVED** [review]

**Pull request merged**